File: medsum_testing/frontend/js/results_actions.js
/**
 * Results-table row actions: Download only.
 * Click a row to open the case. Remove (API) excludes the row from this
 * view; it never deletes Drive, uploads, or results/{id}.json.
 */
(function (root) {
  const ROW_ACTIONS = ['Download'];
  const excludedIds = new Set();

  function stableId(row) {
    if (!row) return '';
    if (root.MedsumTestCaseView && root.MedsumTestCaseView.stableTestId) {
      return String(root.MedsumTestCaseView.stableTestId(row) || '');
    }
    return String(row.test_id || row.id || '');
  }

  function excludeFromView(testId) {
    const id = String(testId || '').trim();
    if (id) excludedIds.add(id);
    return id;
  }

  function isExcluded(testId) {
    return excludedIds.has(String(testId || ''));
  }

  function visibleResults(rows) {
    return (rows || []).filter(row => !isExcluded(stableId(row)));
  }

  function rowActionsHtml(testId, esc) {
    const escape = esc || (s => String(s == null ? '' : s));
    const id = escape(testId || '');
    const disabled = testId ? '' : ' disabled';
    return (
      '<span class="row-actions">'
      + `<button type="button" class="view-btn" data-row-action="download" data-test-id="${id}" title="Download this case report"${disabled}>Download</button>`
      + '</span>'
    );
  }

  const api = {
    ROW_ACTIONS,
    REMOVE_DOES_NOT_DELETE_SOURCE: true,
    excludeFromView,
    isExcluded,
    visibleResults,
    rowActionsHtml,
    stableId,
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api;
  }
  root.MedsumResultsActions = api;
})(typeof window !== 'undefined' ? window : globalThis);

File: medsum_testing/backend/services/results_actions.py
"""Results-table row actions. Remove never deletes source audio or saved JSON.

The only row action is Download. Click a results row to open the case.
Remove (if used via API) excludes the case from the current results view
and from that view's total report. It does not delete Google Drive files,
local uploads, or results/{id}.json.
"""

from __future__ import annotations

ROW_ACTIONS = ("Download",)

REMOVE_DOES_NOT_DELETE_SOURCE = True


def visible_results(rows: list[dict] | None, excluded_ids: set[str] | None) -> list[dict]:
    """Drop excluded ids from the view. Source files stay on disk / Drive."""
    drop = {str(i) for i in (excluded_ids or set()) if i}
    if not drop:
        return list(rows or [])
    out = []
    for row in rows or []:
        tid = str(row.get("test_id") or row.get("id") or "")
        if tid and tid in drop:
            continue
        out.append(row)
    return out
